VoiceController.php

```diff
@@ -1,48 +1,79 @@
-public function voice(Request $request){
+<?php
+
+public function submitVote_NeedsKeepJsonResponse(Request $request) // If, for some reason, the json response needs to keep as it is (needs to work with a old version app, for example). Why "voice"? Isn't this a quizz like thing? Vote typo?
+  {
     $request->validate([
-        'question_id'=>'required|int|exists:questions,id',
-        'value'=>'required|boolean',
+      'question_id' => 'required|int|exists:questions,id',
+      'value' => 'required|boolean',
     ]);
 
-    $question=Question::find($request->post('question_id'));
+    $question = Question::find($request->post('question_id'));
     if (!$question)
-        return response()->json([
-            'status'=>404,
-            'message'=>'not found question ..'
-        ]);
-    if ($question->user_id==auth()->id())
-        return response()->json([
-            'status' => 500,
-            'message' => 'The user is not allowed to vote to your question'
-        ]);
-
-    //check if user voted 
-    $voice=Voice::where([
-        ['user_id','=',auth()->id()],
-        ['question_id','=',$request->post('question_id')]
-    ])->first();
-    if (!is_null($voice)&&$voice->value===$request->post('value')) {
-        return response()->json([
-            'status' => 500,
-            'message' => 'The user is not allowed to vote more than once'
-        ]);
-    }else if (!is_null($voice)&&$voice->value!==$request->post('value')){
-        $voice->update([
-            'value'=>$request->post('value')
-        ]);
-        return response()->json([
-            'status'=>201,
-            'message'=>'update your voice'
-        ]);
+      return response()->json([
+        'status' => 404,
+        'message' => 'Question was not found!'
+      ]);
+    if ($question->user_id == auth()->id()) // I hope this is inside a Middleware that ensures the user is logged in.
+      return response()->json([
+        'status' => 500,
+        'message' => "The user is not allowed to vote on it's own question"
+      ]);
+
+
+    $voice = Voice::firstOrCreate([ // Maybe change model's name? Unless it's actually voice for some reason?
+      'user_id' => auth()->id(),
+      'question_id' => $request->post('question_id')
+    ]);
+
+    if ($voice->wasRecentlyCreated) { // If it was created by CREATE parte of firstOrCREATE
+      return response()->json([
```

Comment:
@valascus good ideas, but I think in this case you save the record without the actual value, so you lose `'value'=>$request->post('value')` from the original code.

Reply:
@valascus ok nvm, the second example has this correct.

Reply:
I have added a commit to solve that.  "Forgot important thing! User needs value! …"